src/garage/torch/algos/dqn.py

```diff
@@ -0,0 +1,285 @@
+                tabular.record('Evaluation/100EpRewardMean',
+                               np.mean(self._episode_reward_mean))
+
+            for _ in range(self._steps_per_epoch):
+                trainer.step_path = trainer.obtain_episodes(trainer.step_itr)
+                if hasattr(self.exploration_policy, 'update'):
+                    self.exploration_policy.update(trainer.step_path)
+
+                self._train_once(trainer.step_itr, trainer.step_path)
+                trainer.step_itr += 1
+
+        return np.mean(last_returns)
+
+    def _train_once(self, itr, episodes):
+        """Perform one iteration of training.
+
+        Args:
+            itr (int): Iteration number.
+            episodes (EpisodeBatch): Batch of episodes.
+
+        """
+        self.replay_buffer.add_episode_batch(episodes)
+
+        epoch = itr / self._steps_per_epoch
+
+        for _ in range(self._n_train_steps):
+            if (self.replay_buffer.n_transitions_stored >=
+                    self._min_buffer_size):
+                timesteps = self.replay_buffer.sample_timesteps(
+                    self._buffer_batch_size)
+                qf_loss, y, q = tuple(v.cpu().numpy()
+                                      for v in self._optimize_qf(timesteps))
+
+                self._episode_qf_losses.append(qf_loss)
+                self._epoch_ys.append(y)
+                self._epoch_qs.append(q)
+
+        if itr % self._steps_per_epoch == 0:
+            self._log_eval_results(epoch)
+
+        if itr % self._target_update_freq == 0:
+            self._target_qf = copy.deepcopy(self._qf)
+
+    def _log_eval_results(self, epoch):
+        """Log evaluation results after an epoch.
+
+        Args:
+            epoch (int): Current epoch.
+        """
+        logger.log('Training finished')
+
+        if self.replay_buffer.n_transitions_stored >= self._min_buffer_size:
+            tabular.record('Epoch', epoch)
+            tabular.record('QFunction/AverageQFunctionLoss',
+                           np.mean(self._episode_qf_losses))
+            tabular.record('QFunction/AverageQ', np.mean(self._epoch_qs))
+            tabular.record('QFunction/MaxQ', np.max(self._epoch_qs))
+            tabular.record('QFunction/AverageAbsQ',
+                           np.mean(np.abs(self._epoch_qs)))
+            tabular.record('QFunction/AverageY', np.mean(self._epoch_ys))
+            tabular.record('QFunction/MaxY', np.max(self._epoch_ys))
+            tabular.record('QFunction/AverageAbsY',
+                           np.mean(np.abs(self._epoch_ys)))
+
+    def _optimize_qf(self, timesteps):
+        """Perform algorithm optimizing.
+
+        Args:
+            timesteps (TimeStepBatch): Processed batch data.
+
+        Returns:
+            qval_loss: Loss of Q-value predicted by the Q-network.
+            ys: y_s.
+            qval: Q-value predicted by the Q-network.
+
+        """
+        observations = np_to_torch(timesteps.observations)
```

Comment:
can you file an issue to add TimeStepBatch.torch()?

Reply:
done